intellij-plugin/src/main/kotlin/com/apollographql/ijplugin/inspection/ApolloMissingGraphQLDefinitionImportInspection.kt

```diff
@@ -0,0 +1,208 @@
+        val argumentTypeToImport = argumentDefinition.type.rawType().name.takeIf { it in knownDefinitionNames } ?: continue
+        if (!isImported(directiveElement, argumentTypeToImport, definitionsUrl)) {
+          val typeKind = getTypeKind(argumentTypeToImport)
+          holder.registerProblem(
+              argument,
+              ApolloBundle.message(message, typeKind, argumentTypeToImport),
+              highlightType,
+              ImportDefinitionQuickFix(typeKind = typeKind, elementName = argumentTypeToImport, definitions = definitions, definitionsUrl = definitionsUrl),
+          )
+        }
+      }
+    }
+  }
+}
+
+private fun getTypeKind(typeName: String): String {
+  val typeDefinition = NULLABILITY_DEFINITIONS.firstOrNull { it is GQLNamed && it.name == typeName } ?: return "unknown"
+  return ApolloBundle.message(
+      when (typeDefinition) {
+        is GQLDirectiveDefinition -> "inspection.missingGraphQLDefinitionImport.reportText.directive"
+        is GQLEnumTypeDefinition -> "inspection.missingGraphQLDefinitionImport.reportText.enum"
+        is GQLInputObjectTypeDefinition -> "inspection.missingGraphQLDefinitionImport.reportText.input"
+        is GQLScalarTypeDefinition -> "inspection.missingGraphQLDefinitionImport.reportText.scalar"
+        else -> return "unknown"
+      }
+  )
+}
+
+private class ImportDefinitionQuickFix(
+    val typeKind: String,
+    val elementName: String,
+    private val definitions: List<GQLDefinition>,
+    private val definitionsUrl: String,
+) : LocalQuickFix {
+  override fun getName() = ApolloBundle.message("inspection.missingGraphQLDefinitionImport.quickFix", typeKind, "'$elementName'")
+  override fun getFamilyName() = name
+
+  override fun availableInBatchMode() = false
+  override fun generatePreview(project: Project, previewDescriptor: ProblemDescriptor): IntentionPreviewInfo = IntentionPreviewInfo.EMPTY
+
+  override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+    if (!IntentionPreviewUtils.isIntentionPreviewActive()) project.telemetryService.logEvent(TelemetryEvent.ApolloIjMissingGraphQLDefinitionImportQuickFix())
+
+    val element = descriptor.psiElement.parentOfType<GraphQLDirective>(withSelf = true)!!
+    val schemaFiles = element.schemaFiles()
+    val linkDirective = schemaFiles.flatMap { it.linkDirectives(definitionsUrl) }.firstOrNull()
+
+    // Special case: also add a @catch directive to the schema if we're importing @catch
+    val catchDirectiveSchemaExtension = if (element.name == CATCH && schemaFiles.flatMap { it.schemaCatchDirectives() }.isEmpty()) {
+      createCatchDirectiveSchemaExtension(project)
+    } else {
+      null
+    }
+
+    if (linkDirective == null) {
+      val linkDirectiveSchemaExtension = createLinkDirectiveSchemaExtension(project, setOf(element.nameForImport), definitions, definitionsUrl)
+      val extraSchemaFile = schemaFiles.firstOrNull { it.name == "extra.graphqls" }
```

Comment:
Should we be more strict in our Gradle API and allow only 2 schema files? (main + extra)

Reply:
Not completely sure - it would make something that has been a convention so far to be more of an "enforced" thing - but there's a risk it will break somebody's project.

Reply:
Yea, in retrospect, should probably have been more strict from day 1. But that ship has sailed